ec/src/hashing/curve_maps/swu/mod.rs

```diff
@@ -0,0 +1,221 @@
+use crate::models::SWModelParameters;
+use ark_ff::{BigInteger, Field, One, PrimeField, SquareRootField, Zero};
+use ark_std::string::ToString;
+use core::marker::PhantomData;
+
+use crate::{
+    hashing::{map_to_curve_hasher::MapToCurve, HashToCurveError},
+    models::short_weierstrass_jacobian::GroupAffine,
+};
+
+/// Trait defining the necessary parameters for the SWU hash-to-curve method
+/// for the curves of Weierstrass form of:
+/// y^2 = x^3 + a*x + b where ab != 0. From [\[WB2019\]]
+///
+/// - [\[WB2019\]] <https://eprint.iacr.org/2019/403>
+pub trait SWUParams: SWModelParameters {
+    /// An element of the base field that is not a square root see \[WB2019, Section 4\].
+    /// It is also convenient to have $g(b/xi * a)$ to be square. In general
+    /// we use a `XI` with low absolute value coefficients when they are
+    /// represented as integers.
+    const XI: Self::BaseField;
+    /// An arbitrary nonsquare conviniently chosen to be a primitve element of the base field
+    const ZETA: Self::BaseField;
+    /// Square root of `THETA = Self::XI/Self::ZETA`.
+    const XI_ON_ZETA_SQRT: Self::BaseField;
+}
+
+/// Represents the SWU hash-to-curve map defined by `P`.
+pub struct SWUMap<P: SWUParams> {
+    curve_params: PhantomData<fn() -> P>,
+}
+
+/// Trait defining a parity method on the Field elements.
+trait ElementParity<F: Field> {
+    fn parity(element: &F) -> bool {
+        element
+            .to_base_prime_field_elements()
+            .next()
+            .unwrap()
+            .into_repr()
+            .is_odd()
+    }
+}
```

Comment:
We can add a blanket implementation for all fields here, so that we don't need to add manual impls downstream.

Reply:
Also, is this the same parity as that suggested in the IETF draft?